src/odemis/driver/xt_client.py

```diff
@@ -2215,3 +2237,74 @@ def _setMultiBeamMode(self, multi_beam_mode):
         #     self.apertureIndex.value = current_aperture
         #     self.beamletIndex.value = current_beamlet
         return (self.parent.get_use_case() == 'MultiBeamTile')
+
+
+class MultiBeamDetector(Detector):
+    """
+    This is an extension of xt_client.Detector class. It overwrites the image acquisition
+    to work with image acquisition in XTToolkit.
+    """
+
+    def __init__(self, name, role, parent, channel_name, **kwargs):
+        """
+        channel_name (str): Name of one of the electron channels.
+        """
+        # The acquisition is based on a FSM that roughly looks like this:
+        # Event\State |    Stopped    |   Acquiring    | Receiving data |
+        #    START    | Ready for acq |        .       |       .        |
+        #    DATA     |       .       | Receiving data |       .        |
+        #    STOP     |       .       |     Stopped    |    Stopped     |
+        #    TERM     |     Final     |      Final     |     Final      |
+
+        Detector.__init__(self, name, role, parent=parent, channel_name=channel_name, **kwargs)
+
+    def stop_generate(self):
+        logging.debug("Stopping image acquisition")
+        # FIXME there has to be a better way to do this :(
+        # This doesn't work:
+        # self.parent.set_channel_state(self._channel_name, False)
+        server = Pyro5.api.Proxy("PYRO:Microscope@192.168.31.158:4242")
+        server.set_channel_state(self._channel_name, False)
```

Comment:
@pieleric I tried stopping the scan by using `self.parent.set_channel_state(self._channel_name, False)`. This just waited for the scan_image call to finish before making the call. 

It has something to do with the `threading.Lock()` we use in the calls to the adapter. Because the `scan_image` call uses the `threading.Lock()` I cannot also run the `set_channel_state` call. Can I somehow unlock the Lock?

As shown on line 2266 and 2267 if I create a new server object I can stop the channel, and stop the scan image call. However this is definitely very hacky.

Reply:
That's one of the main difference between Pyro4 and Pyro5. Pyro4 is able to automatically create a new connection, if the current one is in use.
Definitely, you cannot just "unlock the lock". You could just not use the Lock at all, and use the same connection, but then mayhem would happen ;-)

I'd suggest to create multiple connections, as you have started to do. Here, it's a little hacky because you create the connection on the fly. But you could create a dedicated "cancel_connection", with its own lock, and use it only for cancelling. If we ever need more than 2 connections, probably we'll need to implement something fancy as in Pyro4 which allow you to get the "first connection not in use". Let's hope we don't need to go there though....